lib/benchee/benchmark/measure/memory.ex

```diff
@@ -44,17 +55,22 @@ defmodule Benchee.Benchmark.Measure.Memory do
   defp return_memory({memory_usage, result}) when memory_usage < 0, do: {nil, result}
   defp return_memory({memory_usage, result}), do: {memory_usage, result}
 
-  defp measure_memory(fun, tracer) do
-    word_size = :erlang.system_info(:wordsize)
-    {:garbage_collection_info, heap_before} = Process.info(self(), :garbage_collection_info)
-    result = fun.()
-    {:garbage_collection_info, heap_after} = Process.info(self(), :garbage_collection_info)
-    mem_collected = get_collected_memory(tracer)
+  defp measure_memory(fun, tracer, parent) do
+    :erlang.garbage_collect()
+    send(tracer, :begin_collection)
 
-    memory_used =
-      (total_memory(heap_after) - total_memory(heap_before) + mem_collected) * word_size
+    receive do
+      :ready -> nil
+    end
 
-    {memory_used, result}
+    result = fun.()
+    send(parent, {:result, result})
+    :erlang.garbage_collect()
```

Comment:
do we need to garbage collect here? SHouldn't we be safe without? I'd be worried of the garbage collection induced memory overhead (maybe) although it doesn't really seem to be the case

Reply:
This second garbage collection is mostly used to collect the term bound to the `result` variable on line 66. Without the second collection, we don't count the final return value of the function (and potentially more terms allocated on the heap) towards the total memory used. That's also why we send that result to the parent process before collecting - to be sure the term bound to `result` is free to be GC'd.

Reply:
Huh interesting, I'd think it should show up on the result side because it should be accounted for in the heap memory info shouldn't it? :thinking:

Reply:
It is possible that the final term is allocated in the existing heap after the final (natural) garbage collection, leaving that term on the heap unaccounted for since we're no longer checking the size of the heap with `Process.info/2` after the function we're benchmarking returns. It seemed to happen somewhat frequently, actually, since I noticed significant differences in measurements without this second GC call. Check out what happens without it:

```
Compiling 10 files (.ex)
Excluding tags: [needs_fast_function_repetition: true]

.........................................................................................................................................................................................................................................................................................

  1) test .run_scenarios correctly scales down memory usage of very fast functions (Benchee.Benchmark.RunnerTest)
     test/benchee/benchmark/runner_test.exs:138
     Assertion with >= failed
     code:  assert memory_consumption >= 1
     left:  0
     right: 1
     stacktrace:
       test/benchee/benchmark/runner_test.exs:149: (test)



  2) test .run_scenarios records memory when the function only runs once (Benchee.Benchmark.RunnerTest)
     test/benchee/benchmark/runner_test.exs:122
     Assertion with > failed, both sides are exactly equal
     code: assert length(memory_usages) > 0
     left: 0
     stacktrace:
       test/benchee/benchmark/runner_test.exs:134: (test)

...

  3) test .run_scenarios measures the memory usage of a scenario (Benchee.Benchmark.RunnerTest)
     test/benchee/benchmark/runner_test.exs:105
     Assertion with > failed, both sides are exactly equal
     code: assert length(memory_usages) > 0
     left: 0
     stacktrace:
       test/benchee/benchmark/runner_test.exs:118: (test)

....................................................

  4) test measure/1 returns the result of the function and the memory used (in bytes) (Benchee.MemoryMeasureTest)
     test/benchee/benchmark/measure/memory_test.exs:12
     Assertion with > failed
     code:  assert memory_used > 380
     left:  0
     right: 380
     stacktrace:
       test/benchee/benchmark/measure/memory_test.exs:18: (test)



  5) test measure/1 doesn't return broken values (Benchee.MemoryMeasureTest)
     test/benchee/benchmark/measure/memory_test.exs:22
     Assertion with >= failed
     code:  assert memory_used >= 8000
     left:  5208
     right: 8000
     stacktrace:
       test/benchee/benchmark/measure/memory_test.exs:26: (test)



Finished in 4.5 seconds
110 doctests, 235 tests, 5 failures, 4 excluded

Randomized with seed 865211
```

Reply:
weird - thanks for the background!